lib/Analysis/OptimizeRelinearizationAnalysis/OptimizeRelinearizationAnalysis.h

```diff
@@ -0,0 +1,41 @@
+#ifndef LIB_ANALYSIS_OPTIMIZE_RELINEARIZATIONANALYSIS_H
+#define LIB_ANALYSIS_OPTIMIZE_RELINEARIZATIONANALYSIS_H
+
+#include "llvm/include/llvm/ADT/DenseMap.h"  // from @llvm-project
+#include "mlir/include/mlir/IR/Operation.h"  // from @llvm-project
+#include "mlir/include/mlir/IR/Value.h"      // from @llvm-project
+
+namespace mlir {
+namespace heir {
+class OptimizeRelinearizationAnalysis {
+ public:
+  OptimizeRelinearizationAnalysis(Operation *op) : opToRunOn(op){};
+  ~OptimizeRelinearizationAnalysis() = default;
+
+  LogicalResult solve();
+
+  // Return true if a relin op should be inserted after the given
+  // operation, according to the solution to the optimization problem.
+  bool shouldInsertRelin(Operation *op) const { return solution.lookup(op); }
+
+  // Return the key basis degree at the given SSA value, as determined by the
```

Comment:
maybe ciphertext size or key size (minus 1) ? for rotated ciphertexts, whose basis is `(1, s^2)`, this analysis will still perform correctly since the size of the ciphertexts still increases the same way that `(1, s)` canonical ciphertexts do, although their "key basis degree" is probably greater than 1.

Reply:
This is a point at which my lack of understanding of the cryptosystem implementation is hurting me. We always start at (1, s), and the relin always goes back to (1, s), but under what circumstances do we get other key bases? 

- Multiplication goes (1, s) -> (1, s, s^2), but can also go (1, s, s^2) (1, s, s^2) -> (1, s, s^2, s^3, s^4), right?
- Does rotation go (1, s) -> (1, s^2)?
- Can multiplication handle arguments in different key bases? (1, s) * (1, s^2) for example?

Reply:
* In theory, yes, multiplication can easily be defined for larger key basis, though implementations of that are, afaik, rare
* Rotation does something weirder. There's still a switch of keys involved, but it's somewhat orthogonal to the key basis[^1]. Since rotation virtually always includes keyswitching, you can think of it as a no-op on the key basis (1,s) in, (1,s) out.
* I'm not actually sure about that, never thought about it. Given that higher-than-(1,s)-mul is already super rare, I guess it might not matter in practice.



[^1]: I think it's something like permuting coefficients but I'm pretty rusty on the details here. I believe one could think of rotation (w/o keyswitching, which is, other than for mul, virtually always considered an implicit aprt of the operation) as going form (1, s) to (1, s') where s' is some "permuted" version).

Reply:
> Multiplication goes (1, s) -> (1, s, s^2), but can also go (1, s, s^2) (1, s, s^2) -> (1, s, s^2, s^3, s^4), right?

Correct! Although not many libraries will implement that high dimension multiplication (so I think your max basis constant is useful)

> Does rotation go (1, s) -> (1, s^2)?

Or (1, s^i) -> (1, s^j) more generally. In theory it could also rotate (1, s, s^2) to (1, s^i, s^2^i), but again i doubt anyone is exposing that.

> Can multiplication handle arguments in different key bases? (1, s) * (1, s^2) for example?

Yes, in theory, but again not exposed. I worked it out on paper and unfortunately a mul between (1, s^i) and (1, s^j) will end up being a result of ciphertext with basis (1, s^i, s^j, s^{i+j}). In the case where i and j agree is where we see the subtraction by one since the cross terms combine to (1, s^i, s^{i + i})

Reply:
> > Does rotation go (1, s) -> (1, s^2)?
> 
> Or (1, s^i) -> (1, s^j) more generally. In theory it could also rotate (1, s, s^2) to (1, s^i, s^2^i), but again i doubt anyone is exposing that.

can you go into more detail on that one? I thought the key switch that happened during rotations was orthogonal to the degrees, but I guess I'm mistaken on that?

Reply:
> can you go into more detail on that one? I thought the key switch that happened during rotations was orthogonal to the degrees, but I guess I'm mistaken on that?

Yeah, the key switch that happens during rotations is orthogonal to the key switching that changes a ciphertext key basis from `(1, s, s^2)` to `(1, s)`. But actually key switching keys are all generated from the same principle. I'm not sure which part to go into, but Galois rotations are just performed by applying the automorphism `X -> X^i` on the polynomial. So that can be applied to a size 3 ciphertext or a size two ciphertext. The same automorphism will apply to the secret key it is encrypted under (so you would map (1, s) to (1, s^i)).

Reply:
Here's an implementation in BGV: https://eprint.iacr.org/2011/566.pdf
The key switching procedure on page 29 can be done from any type of key, like from (1, s^i) back to (1, s) or from (1, s, s^2) back to (1, s) or from (1, s^i, s^2^i) to (1, s).

Obviously the client needs to generate all of these though.

Reply:
Ohh, I see now what you mean - that's also my understanding of rotate and keyswitch, but I think it's still better to say that "rotation does not / cannot modify the size of the key basis", and to instead talk about (generic) keyswitching separately:  "could also ~~rotate~~ keyswitch (1, s, s^2) to (1, s^i, s^2^i)".

In general, it seems like the (applied) FHE community has slightly inconsistent terminologoy:
`rotate` usually meansto "automorphism + keyswitch" where as  `mul` generally only means "tensor-like-product", and the associated keyswitch is a  separate `relin` operation. (The crypto-y  view of FHE is a bit more consistent, and calls what most libraries call multiply the "pre-multiply", with "multiply" defined as including the keyswitch/relin)

Reply:
> actually, i did mean rotate - that doesn't change the size either. that meant just permuting the coefficients with the automorphism

Now I'm confused... let me try and make the keybasis notation a bit more explicit:
In my understanding, a rotate-by-two (w/o keyswitch) gives you a change from (1, s(X)) -> (1, s(X^2)) whereas the key basis change in a multiply (w/o keyswitch/relin) is (1, s(X)) -> (1, s(X), s(X)^2)) which is fundamentally different?

Reply:
They are! Now take a multiplied ciphertext with basis (1, s(X), s(X)^2) and then apply the X -> X^2 transformation and now you've transformed from (1, S(X), S(X)^2) to (1, S(X^i), S(X^i)^2) (the transform in that comment)

Reply:
I'm trying to figure out what simplification works best for now. Maybe assume rotation is always linear in/out, and constrain it appropriately, and so we always have degree == dimension-1. That would require no changes to this PR, but maybe some additional comments explaining the assumptions

Reply:
I think that's reasonable. Most libraries will always assume rotations expect (1, s) in